### PR TITLE
Add method to check if point is inside cell

### PR DESCRIPTION
**Pull request by MatthewFlamm**
### Overview

Add method to check if point is inside a specific cell.  This was motivated partly by the discussion in https://github.com/pyvista/pyvista/discussions/2703#discussioncomment-2828184

### Details

Related to #2897.

Should we allow a Sequence of points (if more than one is provided we can return a numpy ndarray)?  If so, this might complicate any plans to allow for Iterables, Sequences, or similar of cell indices.

### CODE REVIEW

File: pyvista/core/dataset.py
Comment (tkoyama010) [inline pyvista/core/dataset.py:2528]:
```suggestion
    def point_is_inside_cell(self, ind: int, point: Vector) -> bool:
```
I guess this is more clear what the method is doing.

File: pyvista/core/dataset.py
Comment (tkoyama010) [inline pyvista/core/dataset.py:2534]: Is this Cell ID ?
```suggestion
            Cell ID.
```

**Review by tkoyama010 — CHANGES_REQUESTED** [review]
Thanks for adding a useful method.

File: pyvista/core/dataset.py
Comment (MatthewFlamm) [reply pyvista/core/dataset.py]: Thanks, this is indeed better.  I was trying to be thematically similar to the `DataSet.cell_*` methods like `DataSet.cell_bounds` as it indicates it is operating on a single cell basis.  But I think you name is the most clear.  We can have a `See Also` section potentially to address my thinking.

I will clean up the usage after committing your suggestion directly.

**Comment by MatthewFlamm** [comment]
Unlike `DataSet.find_containing_cell`, there is only minor efficiency gained by allowing multiple points to be supplied here.  The inefficiency here is multiple calls to `GetCell(ind)` and `GetPoints().GetNumberOfPoints()` (I'm assuming the calls to `vtk.mutable` are cheap), which would be avoided if we allowed multiple points to be supplied to this function.  Should we implement multiple points here?  I'm not sure based on some discussions about also supporting multiple cells here https://github.com/pyvista/pyvista/pull/2897#discussion_r908767542.  If we support multiple points and multiple cells, then it becomes unclear, at least to me, what the 2D output structure should look like.

cc @adeak  and @banesullivan

**Review by tkoyama010 — APPROVED** [review]
We need to wait for a reply to the above, otherwise, it is LGTM.

File: pyvista/core/dataset.py
Comment (adeak) [inline pyvista/core/dataset.py:2552]: Nitpick:
```suggestion
        if not 0 <= ind < self.n_cells:
```
at least I always find it easier to reason about a single interval, despite the negation.

File: pyvista/core/dataset.py
Comment (adeak) [inline pyvista/core/dataset.py:2559]: One pro for allowing multiple points: we could use your `_coerce_pointslike_arg()` to do input validation.

File: pyvista/core/dataset.py
Comment (adeak) [inline pyvista/core/dataset.py:2567]: If we don't use these mutables after the call, do they still have to be mutable? Do we get a segfault otherwise?

**Review by adeak — COMMENTED** [review]
> Unlike `DataSet.find_containing_cell`, there is only minor efficiency gained by allowing multiple points to be supplied here. The inefficiency here is multiple calls to `GetCell(ind)` and `GetPoints().GetNumberOfPoints()` (I'm assuming the calls to `vtk.mutable` are cheap), which would be avoided if we allowed multiple points to be supplied to this function. Should we implement multiple points here? I'm not sure based on some discussions about also supporting multiple cells here [#2897 (comment)](https://github.com/pyvista/pyvista/pull/2897#discussion_r908767542). If we support multiple points and multiple cells, then it becomes unclear, at least to me, what the 2D output structure should look like.

My main reason for contemplating multiple points/cells as input is not that of performance (I don't typically associate VTK with vectorization), but rather convenience. If one has more than one point to check it would be nicer to call the method once with every point, rather than using something like a list comprehension outside the function, point by point.

This ties in with your question: if the API becomes confusing in the "multiple points, multiple cells" case then we should just forbid it. It's not convenient to have to look up the docs to see how it behaves.

That being said, I could imagine that if either points or cells are multiple then we return an array of bools (as large as the number of input points or cells), and if both are sequences then we return a 2d array of shape `(n_points, n_indices)` or `(n_indices, n_points)`. But again, this might be too smart for its own good.

**Comment by MatthewFlamm** [comment]
I agree on the convenience factor when using one or the other.  I find the convenience factor will drop when using both together.  I will never remember which comes first in the indices.  But, maybe this is a rare enough use case that it isn't worth trying to over optimize the API.  I'll implement multiple points

File: pyvista/core/dataset.py
Comment (MatthewFlamm) [reply pyvista/core/dataset.py:2567]: A bit better than a segfault, but a `TypeError`:

```
E   TypeError: EvaluatePosition argument 3: a vtk.reference() object is required
```

**Review by tkoyama010 — APPROVED** [review]

File: pyvista/core/dataset.py
Comment (adeak) [inline pyvista/core/dataset.py:2575]: In the edge case of someone passing in a `(1, 3)`-shaped arraylike, is it OK that this will return `True` instead of `[True]`? If the API says "return a scalar for a scalar", it might be more consistent to respect if the input is `[(0, 0, 0)]` instead of `(0, 0, 0)`. But unpacking the singular case is fine too, just something to consider.

File: pyvista/core/dataset.py
Comment (adeak) [inline pyvista/core/dataset.py:2557]: Should we add a `TypeError` for non-integer `ind`? I don't insist on doing that since most weird cases will fail loudly with this check. Only the "passed in a float by accident" case might be relevant for this, assuming that such accidental floats are not int-valued floats.

File: pyvista/core/dataset.py
Comment (MatthewFlamm) [reply pyvista/core/dataset.py:2575]: Good point.  We currently have this problem in the other functions too: 

https://github.com/pyvista/pyvista/blob/f8208ad25c518d3338e57a821a9bd0772286b08b/pyvista/core/dataset.py#L2285

Comment (MatthewFlamm) [reply pyvista/core/dataset.py:2575]: I think this is important to fix.  A user might be dynamically generating a list of points to use, it will be unexpected that they should have to check for arrays of shape (1, 3).

Since _points_like_arg is a private method, we have free reign to change it.  I'm not sure if we should take extra steps to warn users that the behavior has changed, or not.  I argue that the current behavior is a bug, so we can just change it.

Comment (adeak) [reply pyvista/core/dataset.py:2578]: Although fixing the bug mostly means that some ifs in downstream code will merely be redundant, we should probably annotate the involved public functions with `.. versionadded:`. Otherwise people may never notice.

Comment (adeak) [reply pyvista/core/dataset.py:2578]: If you meant emitting warnings: it could go either way. Not strictly necessary, but this is a very high signal to noise ratio warning.

Comment (MatthewFlamm) [reply pyvista/core/dataset.py:2578]: > If you meant emitting warnings: it could go either way. Not strictly necessary, but this would be a very high signal to noise ratio warning.

One issue is that if we emit a warning in `_coerce_pointslike_arg` for an array of shape `(1, 3)` to warn users of the behavior change, it means that even after the user adjusts their downstream code, it will still emit the warning.  I like the approach of adding to the docstring a clear description of the change, so users can figure out what happened.

Comment (adeak) [reply pyvista/core/dataset.py:2578]: Yeah, I realized that too eventually. It would be useful the first time but useless afterward. Agreed on the doc annotation.

File: pyvista/core/dataset.py
Comment (MatthewFlamm) [inline pyvista/core/dataset.py:2304]:
```suggestion
```

File: pyvista/utilities/common.py
Comment (MatthewFlamm) [inline pyvista/utilities/common.py:31]:
```suggestion

```

File: pyvista/core/dataset.py
Comment (MatthewFlamm) [reply pyvista/core/dataset.py]: I used `..versionchanged:: 0.35.0`, but I'm not 100% confident that this is formatted correctly in the right place.

File: pyvista/utilities/common.py
Comment (adeak) [inline pyvista/utilities/common.py:30]: This is definitely subjective, but I'd prefer to be exact about the distinction, because it should be unambiguous. Something like
```suggestion
        Whether the input was a single point. This excludes a 2d array-like with
        shape ``(1, 3)``.
```

Or may be go even more direct and just say "Whether the input was a single point in an array-like with shape `(3,)`".

File: pyvista/utilities/common.py
Comment (adeak) [inline pyvista/utilities/common.py:26]: I don't know how consistent the codebase is in this, but there's precedent of keeping only types here even if there are multiple return values, e.g. https://github.com/pyvista/pyvista/blob/709293784756e7f0fb4f4125a95fafc02eab0da5/pyvista/core/filters/poly_data.py#L3181-L3189

File: pyvista/core/dataset.py
Comment (adeak) [inline pyvista/core/dataset.py:2191]:
```suggestion
               Inputs of shape ``(1, 3)`` now return a :class:`numpy.ndarray`
               of shape ``(1,)``.
```

File: pyvista/core/dataset.py
Comment (adeak) [inline pyvista/core/dataset.py:2199]:
```suggestion
               Inputs of shape ``(1, 3)`` now return a :class:`numpy.ndarray`
               of the same shape.
```

File: pyvista/core/dataset.py
Comment (adeak) [inline pyvista/core/dataset.py:2317]:
```suggestion
               Inputs of shape ``(1, 3)`` now return a :class:`numpy.ndarray`
               of shape ``(1,)``.
```

File: pyvista/core/dataset.py
Comment (adeak) [inline pyvista/core/dataset.py:2582]:
```suggestion
        """Return whether one or more points are inside a cell.
```

File: pyvista/core/dataset.py
Comment (adeak) [inline pyvista/core/dataset.py:2580]: How strong feelings do you have about having the cell index first and the point(s) second? I'm wondering if we should flip the order. Or are there other similar filters that this might match?

File: pyvista/core/dataset.py
Comment (adeak) [inline pyvista/core/dataset.py:2591]: I don't really know the specification of these fields, but how standard is `Sequence(float)`? I can only see this in `_coerce_pointslike_arg()` and `find_containing_cell()`, `find_closest_cell()`. If it's like `typing.Sequence` then perhaps it should be `Sequence[float]`?

OK, we have both in https://dev.pyvista.org/api/core/_autosummary/pyvista.DataSet.find_closest_cell.html?highlight=find_closest_cell#pyvista.DataSet.find_closest_cell vs https://dev.pyvista.org/api/core/_autosummary/pyvista.core.common_data.perlin_noise.html?highlight=perlin_noise, and neither get picked up by intersphinx so it probably doesn't really matter either way.

File: pyvista/core/dataset.py
Comment (adeak) [inline pyvista/core/dataset.py:2598]:
```suggestion
            Whether point(s) is/are inside cell. A scalar bool is only returned if
            the input point has shape ``(3,)``.
```

File: pyvista/core/dataset.py
Comment (adeak) [inline pyvista/core/dataset.py:2610]: Sorry to keep bikeshedding this, but numpy ints might pop up their head. What do you think?
```suggestion
        if not isinstance(ind, (int, np.integer)):
```

File: pyvista/core/dataset.py
Comment (adeak) [inline pyvista/core/dataset.py:2629]: [VTK docs](https://vtk.org/doc/nightly/html/classvtkCell.html#a188d8c01ab1cc03e1fd8b04e3ec97faf) says this can return -1 in case of a "computational problem". We might have to pull out the `bool` conversion, and raise/warn-and-set-to-NaN any broken values. Although the NaN setting would be a but problematic if we're aiming for a bool array.

File: pyvista/core/dataset.py
Comment (adeak) [reply pyvista/core/dataset.py]: I see you've pushed a corresponding chance since. Looks good to me now, resolving.

File: pyvista/utilities/common.py
Comment (akaszynski) [reply pyvista/utilities/common.py:26]: That's the precedent, but we can change this if we feel that it needs to be changed. numpydoc specifies that this is fine.

Personally I like how clean it looks without the variable return name, and it also makes maintenance a bit easier.

File: tests/utilities/test_common.py
Comment (MatthewFlamm) [reply tests/utilities/test_common.py:31]: I don't have a strong opinion on this.  My thought was the above usages cannot use the variable, so I chose to be consistent, but I can see the use of the variable here too.

File: pyvista/utilities/common.py
Comment (MatthewFlamm) [reply pyvista/utilities/common.py]: Took your later suggestion.  It is the most direct.  This is also a private helper, so direct is best IMO.

File: pyvista/utilities/common.py
Comment (MatthewFlamm) [reply pyvista/utilities/common.py]: Three non-conforming cases are probably all committed by me.  Sorry.  It is my personal preference, but I should be more careful with the pyvista style.

File: pyvista/core/dataset.py
Comment (MatthewFlamm) [inline pyvista/core/dataset.py:2594]:
```suggestion
        point : Sequence[float] or np.ndarray
```

File: pyvista/core/dataset.py
Comment (MatthewFlamm) [reply pyvista/core/dataset.py:2633]: Good catch, let's check and raise if not 0 or 1.  It makes sense to me to have a bool (array) return for this.

File: pyvista/core/dataset.py
Comment (MatthewFlamm) [reply pyvista/core/dataset.py]: Went with square brackets.

File: pyvista/core/dataset.py
Comment (adeak) [reply pyvista/core/dataset.py:2633]: I'd prefer raising too, my only concern is what happens if someone passes 100 points and one of them has a computational problem? They'd have to remove those points somehow and use the rest.

Comment (MatthewFlamm) [reply pyvista/core/dataset.py:2633]: I suppose we should keep the vtk API but it doesn't seem intuitive.    One thought is to put in a Raises section and to suggest to the user to not supply a set of points if they want to catch an error for single points.

Comment (adeak) [reply pyvista/core/dataset.py:2633]: Hmm, yeah, I think that's a ~good~ least bad compromise. I would also love to keep the bool dtype and the raising. Hopefully computational problems are rare enough that this should only affect few people in very rare cases. We could also mention the recommendation in the error message.

In the off-chance that this starts causing issues downstream we could probably switch from bool to float without breaking existing code.

Comment (MatthewFlamm) [reply pyvista/core/dataset.py:2633]: I suspect this will be a very rare corner case, possibly related to parametric cells (I don't know too much about these, but there are several caveats about them in the docs).  vtk docs also mention that this is rare, and there is a very workable solution for those cases, i.e. loop point by point, which shouldn't be that much slower.

File: tests/utilities/test_common.py
Comment (adeak) [reply tests/utilities/test_common.py:31]: Hmm. If I understand your point correctly, you could use `np.array_equal(coerced_arg, [point])` for the earlier cases. Would that help?

File: pyvista/core/dataset.py
Comment (MatthewFlamm) [reply pyvista/core/dataset.py:2583]: I chose this since this is a cell wise operation, which feels like it should be the first parameter to me, but I'm open to other opinions.  I like the idea of an eventual API like `dataset.cells[10].is_point_inside(points)`, which would suggest the current order for now.  But there was also the suggestion of being able to provide a Sequence or similar of cells to this type of function.

File: pyvista/core/dataset.py
Comment (MatthewFlamm) [reply pyvista/core/dataset.py]: Implemented a loop to catch `-1` or any other unexpected value.

File: tests/utilities/test_common.py
Comment (MatthewFlamm) [reply tests/utilities/test_common.py:31]: On second consideration, I slightly prefer explicitness here.  The data isn't copied sometimes, so we could modify `point` sometimes.

**Comment by akaszynski** [comment]
Since 0.35 will be released soon, I'd like this to wait for the next minor (0.36) release. Best to not add any last second features.

Good news is we won't wait 3 months to get out the next release. There are a few critical features that we need much earlier, so expect 0.36.0 to be out in a month.

**Comment by adeak** [comment]
Oh, I forgot that this was waiting for me, I'm sorry @MatthewFlamm.

@akaszynski I think this is be good to go, if you can give me ten minutes to make sure.

File: pyvista/core/dataset.py
Comment (adeak) [inline pyvista/core/dataset.py:2636]: Two cosmetic remarks here that we could do:
1. skip coverage for the esoteric error case to make coverage happy
2. also add the point index `i` in the error.

I find it much more valuable to get this feature out in v0.35 than to fix this "should almost never happen" edge case and add a doc-build-length delay. We can follow up for the subsequent release if we want to.

Approving, thanks and sorry for forgetting about this.

File: pyvista/core/dataset.py
Comment (adeak) [reply pyvista/core/dataset.py:2583]: OK, let's leave this as is.

**Review by adeak — APPROVED** [review]

**Pull request merged**
